src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1099,18 +1104,20 @@ uct_ud_ep_do_pending(ucs_arbiter_t *arbiter, ucs_arbiter_elem_t *elem,
      * - not in async progress
      * - there are only low priority ctl pending or not ctl at all
      */
-    if (!in_async_progress &&
+    if ((!in_async_progress || (uct_ud_pending_req_priv(req)->flags &
```

Comment:
can we add new local variable like "allow_callback" to be `(!in_async_progress || (uct_ud_pending_req_priv(req)->flags &  UCT_CB_FLAG_ASYNC))`